File: mobile/lib/src/finalytics/finalytics_service.dart
import 'dart:math';

import 'package:finny/src/accounts/accounts_service.dart';
import 'package:finny/src/goals/goal_model.dart';
import 'package:finny/src/goals/goals_service.dart';
import 'package:finny/src/powersync/database.dart';
import 'package:finny/src/profile/profile_service.dart';

class FinalyticsService {
  final AppDatabase appDb;
  late final GoalsService goalsService;
  late final ProfileService profileService;

  FinalyticsService({required this.appDb}) {
    goalsService = GoalsService(
        appDb: appDb, accountsService: AccountsService(appDb: appDb));
    profileService = ProfileService(appDb: appDb);
  }

  Future<int> getActualRetirementAge() async {
    final fv = await getFreedomFutureValueOfCurrentExpenses();
    final pv = await getAssignedBalanceOnRetirementGoal();
    final pmt = await getActualSavingsThisMonthForRetirementGoal() * 12;
    const interest = 0.08;
    final currentAge = (await profileService.getProfile()).age!;

    final n = calculatePeriodFromFutureValue(
      fv: fv,
      pv: pv,
      payment: pmt,
      annualInterestRate: interest,
    );

    return n + currentAge;
  }

  // Calculate the total amount of money (future value) I need today to sustain my expenses in perpetuity at 4% interest
  Future<double> getFreedomFutureValueOfCurrentExpenses() async {
    final averageMonthlyInflowOutflow = await getAverageMonthlyInflowOutflow();
    return averageMonthlyInflowOutflow.outflows * 12 / 0.04;
  }

  Future<double> getFreedomFutureValueOfCurrentExpensesAtRetirement() async {
    final freedomFVCurrentAge = await getFreedomFutureValueOfCurrentExpenses();
    final yearsToRetirement = await calculateYearsToRetirement();

    return calculateFutureValue(
        presentValue: freedomFVCurrentAge.abs(),
        yearlyInvestment: 0,
        annualInterestRate: 0.02, // 2% inflation rate
        years: yearsToRetirement);
  }

  Future<double> getTargetSavingsAndInvestmentsThisMonth() async {
    final pv = -((await getAssignedBalanceOnRetirementGoal()).abs());
    final period = await calculateYearsToRetirement();
    const interest = 8.0;
    final fv = await getFreedomFutureValueOfCurrentExpensesAtRetirement();
    final yearlySavingsTarget = calculatePaymentFromFutureValue(
      fv: fv,
      pv: pv,
      interest: interest,
      years: period,
    );

    return yearlySavingsTarget / 12;
  }

  Future<double> getActualSavingsAndInvestmentsThisMonth() async {
    final savings = await getActualSavingsThisMonthForRetirementGoal();
    final investment = await getActualInvestmentThisMonthForRetirementGoal();
    return savings + investment;
  }

  Future<double> getActualSavingsAtRetirement() async {
    const interestRate = 0.08;
    final period = await calculateYearsToRetirement();
    final pv = -((await getAssignedBalanceOnRetirementGoal()).abs());
    final pmt =
        -((await getActualSavingsThisMonthForRetirementGoal()).abs() * 12);

    return calculateFutureValue(
        presentValue: pv,
        yearlyInvestment: pmt,
        annualInterestRate: interestRate,
        years: period);
  }

  Future<double> getActualSavingsThisMonthForRetirementGoal() async {
    final result = await appDb.customSelect('''
      WITH retirement_goal AS (
        SELECT
          id
        FROM
          goals
        WHERE
          goals.goal_type = 'retirement'
        LIMIT 1
      ),
      assigned_accounts AS (
        SELECT
          account_id
        FROM
          goal_accounts
          JOIN retirement_goal ON retirement_goal.id = goal_accounts.goal_id
        WHERE
          goal_accounts.goal_id = retirement_goal.id
      ),
      start_of_month_balances AS (
        -- Get the earliest balance in the current month for each account
        SELECT
          ab.account_id,
          ab.balance_date AS start_balance_date,
          ab.current_balance AS start_balance
        FROM
          account_balances ab
        WHERE
          ab.balance_date = (
            SELECT
              min(balance_date)
            FROM
              account_balances
            WHERE
              account_id = ab.account_id
              AND balance_date >= date('now', 'start of month') -- Start of the current month in SQLite
          ) 
          AND ab.account_id IN (SELECT account_id FROM assigned_accounts)
      ),
      most_recent_balances AS (
        -- Get the most recent balance for each account
        SELECT
          ab.account_id,
          ab.balance_date AS most_recent_balance_date,
          ab.current_balance AS most_recent_balance
        FROM
          account_balances ab
        WHERE
          ab.balance_date = (
            SELECT
              max(balance_date)
            FROM
              account_balances
            WHERE
              account_id = ab.account_id
          )
          AND ab.account_id IN (SELECT account_id FROM assigned_accounts)
      )
      SELECT
        mr.account_id,
        (mr.most_recent_balance - som.start_balance) AS net_balance_change
      FROM
        most_recent_balances mr
        JOIN start_of_month_balances som ON mr.account_id = som.account_id
        JOIN accounts ON mr.account_id = accounts.id
      WHERE
        mr.most_recent_balance_date >= som.start_balance_date;
        ''').get();

    return result.fold<double>(
      0.0,
      (sum, row) => sum + row.read<double>('net_balance_change'),
    );
  }

  Future<double> getActualInvestmentThisMonthForRetirementGoal() async {
    final result = await appDb.customSelect('''
        WITH start_of_month AS (
            -- Get the earliest holding in the current month for each account and security
            SELECT
                ihd.account_id,
                ihd.investment_security_id,
                ihd.holding_date AS earliest_holding_date,
                ihd.quantity AS start_quantity,
                ihd.institution_price AS start_price
            FROM investment_holdings_daily ihd
            WHERE ihd.holding_date = (
                SELECT MIN(holding_date)
                FROM investment_holdings_daily
                WHERE account_id = ihd.account_id
                  AND investment_security_id = ihd.investment_security_id
                  AND holding_date >= DATE('now', 'start of month')  -- Start of the current month
            )
        ),
        most_recent_holdings AS (
            -- Get the most recent holdings for each account and security
            SELECT
                ihd.account_id,
                ihd.investment_security_id,
                ihd.holding_date AS most_recent_holding_date,
                ihd.quantity AS current_quantity,
                ihd.institution_price AS current_price
            FROM investment_holdings_daily ihd
            WHERE ihd.holding_date = (
                SELECT MAX(holding_date)
                FROM investment_holdings_daily
                WHERE account_id = ihd.account_id
                  AND investment_security_id = ihd.investment_security_id
            )
        )
        SELECT
            mrh.account_id,
            mrh.investment_security_id,
            (mrh.current_quantity - som.start_quantity) * mrh.current_price AS personal_investment
        FROM
            most_recent_holdings mrh
        JOIN
            start_of_month som
            ON mrh.account_id = som.account_id
            AND mrh.investment_security_id = som.investment_security_id
        WHERE
            mrh.current_quantity > som.start_quantity  -- Only consider increases in quantity (personal investment)
            AND mrh.most_recent_holding_date >= som.earliest_holding_date;  -- Compare from the earliest date in the current month
    ''').get();

    final totalPersonalInvestment = result.fold<double>(
      0.0,
      (sum, row) => sum + row.read<double>('personal_investment'),
    );

    return totalPersonalInvestment;
  }

  Future<double> getCurrentTotalInvestmentBalance() async {
    final result = await appDb.customSelect('''
      SELECT
        sum(current_balance) as current_investment_balance
      FROM
        accounts
      WHERE
        type = 'investment';
    ''').getSingle();

    return result.read<double>('current_investment_balance');
  }

  /// Get the average monthly inflow and outflow for the last 12 months
  Future<AverageMonthlyInflowOutflow> getAverageMonthlyInflowOutflow() async {
    final result = await appDb.customSelect('''
      WITH monthly_totals AS (
        SELECT
          strftime('%Y-%m', date) AS month,
          SUM(CASE WHEN amount < 0 THEN ABS(amount) ELSE 0 END) AS monthly_inflows,
          SUM(CASE WHEN amount > 0 THEN amount ELSE 0 END) AS monthly_outflows
        FROM
          transactions
          JOIN accounts ON transactions.account_id = accounts.id
        WHERE
          date >= date('now', '-12 months')
        GROUP BY
          month
      )
      SELECT
        AVG(monthly_inflows) AS average_inflows,
        AVG(monthly_outflows) AS average_outflows
      FROM
        monthly_totals;
    ''').getSingle();

    return AverageMonthlyInflowOutflow(
        inflows: result.read<double>('average_inflows'),
        outflows: result.read<double>('average_outflows'));
  }

  Future<int> calculateYearsToRetirement() async {
    final profile = await profileService.getProfile();

    if (profile.age == null) {
      throw Exception('Age is not set');
    }

    if (profile.retirementAge == null) {
      throw Exception('Retirement age is not set');
    }

    return profile.retirementAge! - profile.age!;
  }

  Future<double> getAssignedBalanceOnRetirementGoal() async {
    final retirementGoal = (await goalsService.getGoals()).firstWhere(
      (goal) => goal.goalType == GoalType.retirement,
      orElse: () => throw Exception('No retirement goal found'),
    );

    return (await goalsService.getTotalBalanceAssignedGoals(retirementGoal.id))
        .abs();
  }

  /// Calculate future value with both present value and monthly contributions
  double calculateFutureValue({
    required double presentValue,
    required double yearlyInvestment,
    required double annualInterestRate,
    required int years,
  }) {
    double futureValue = presentValue * pow(1 + annualInterestRate, years) +
        yearlyInvestment *
            ((pow(1 + annualInterestRate, years) - 1) / annualInterestRate);

    if (yearlyInvestment < 0 && presentValue < 0) {
      futureValue = -futureValue;
    }

    return futureValue;
  }

  double calculatePaymentFromFutureValue({
    required double fv,
    required double pv,
    required double interest,
    required int years,
  }) {
    // Convert interest rate to decimal
    double rate = interest / 100;

    // Calculate the payment using the future value formula for annual payments
    double payment;
    if (rate == 0) {
      // If interest rate is 0, use simple division
      payment = (fv - pv) / years;
    } else {
      payment = (fv - pv * pow(1 + rate, years)) /
          ((pow(1 + rate, years) - 1) / rate);
    }

    // Adjust the sign if both FV and PV are negative
    if (fv < 0 && pv < 0) {
      payment = -payment;
    }

    // Round to 2 decimal places
    return double.parse(payment.toStringAsFixed(2));
  }

  int calculatePeriodFromFutureValue({
    required double fv,
    required double pv,
    required double payment,
    required double annualInterestRate,
  }) {
    if (annualInterestRate == 0) {
      // If interest rate is 0, use simple division
      return ((fv - pv) / payment).round();
    }

    fv = fv.abs();
    pv = pv.abs();
    payment = payment.abs();

    double n = log((fv * annualInterestRate + payment) /
            (pv * annualInterestRate + payment)) /
        log(1 + annualInterestRate);

    return n.round();
  }
}

class AverageMonthlyInflowOutflow {
  final double inflows;
  final double outflows;

  AverageMonthlyInflowOutflow({required this.inflows, required this.outflows});
}

File: finny-api/src/api/services/FinalyticsService.scala
package api.services

import api.common.*
import scalasql.*
import scalasql.core.*
import scalasql.core.SqlStr.*

import java.util.UUID
import scala.util.Try

import PostgresDialect.*
import api.models.UserId
import api.models.FinalyticKeys
import api.repositories.ProfileRepository
import api.repositories.GoalRepository
import api.repositories.GoalRepository.getAssignedBalanceOnRetirementGoal

case class FinalyticsTable[T[_]](
    id: T[UUID],
    user_id: T[UUID],
    key: T[String],
    value: T[String]
)

object FinalyticsTable extends Table[FinalyticsTable]() {
  override def tableName: String = "finalytics"
}

enum ExpenseCalculation:
  case Last12Months, Average

object FinalyticsService:
  def getActualRetirementAge(userId: UserId, expCalc: ExpenseCalculation)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Int] =
    val annualInterestRate = 0.08
    for
      profile <- ProfileRepository.getProfile(userId)
      fv <- getFreedomFutureValueOfCurrentExpenses(expCalc, userId)
      pv <- GoalRepository.getAssignedBalanceOnRetirementGoal(userId)
      pmt <- calculateActualSavingsThisMonth(userId)
      currentAge <- Right(profile.age.getOrElse(0))
    yield calculatePeriodFromFutureValue(fv, pv, pmt, annualInterestRate)

  end getActualRetirementAge

  /** Recalculates the actual savings and investments this month and persists
    * the value in the database
    */
  def recalculateActualSavingsAndInvestmentsThisMonth(userId: UserId)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Double] =
    for
      value <- calculateActualSavingsThisMonth(userId)
      _ <- updateFinalytics(
        userId,
        FinalyticKeys.ActualSavingsAndInvestmentsThisMonth,
        value.toString()
      )
    yield value
  end recalculateActualSavingsAndInvestmentsThisMonth

  def recalculateTargetSavingsAndInvestmentsThisMonth(userId: UserId)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Double] =
    for
      value <- calculateTargetSavingsAndInvestmentsThisMonth(userId)
      _ <- updateFinalytics(
        userId,
        FinalyticKeys.TargetSavingsAndInvestmentsThisMonth,
        value.toString()
      )
    yield value
  end recalculateTargetSavingsAndInvestmentsThisMonth

  def recalculateActualSavingsAtRetirement(userId: UserId)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Double] =
    for
      value <- calculateActualSavingsAtRetirement(
        userId,
        ExpenseCalculation.Average
      )
      _ <- updateFinalytics(
        userId,
        FinalyticKeys.ActualSavingsAtRetirement,
        value.toString()
      )
    yield value
  end recalculateActualSavingsAtRetirement

  def getFinalytics(userId: UserId, key: FinalyticKeys)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, String] =
    val query = FinalyticsTable.select
      .filter(r =>
        r.user_id === UserId.toUUID(userId) && r.key === key.toString()
      )
      .single

    val result = Try:
      dbClient.transaction: db =>
        db.run(query)

    result.toEither.left
      .map(e => AppError.DatabaseError(e.getMessage))
      .map(r => r.value)

  def updateFinalytics(
      userId: UserId,
      key: FinalyticKeys,
      value: String
  )(using dbClient: DbClient.DataSource): Either[AppError, Int] =
    val query = FinalyticsTable.insert
      .columns(
        _.user_id := UserId.toUUID(userId),
        _.key := key.toString(),
        _.value := value
      )
      .onConflictUpdate(_.user_id, _.key)(_.value := value)

    val result = Try:
      dbClient.transaction: db =>
        db.run(query)

    result.toEither.left
      .map(e => AppError.DatabaseError(e.getMessage))
      .map(rows => rows)

  def calculateActualSavingsThisMonth(
      userId: UUID
  )(using dbClient: DbClient.DataSource): Either[AppError, Double] =
    Try:
      val result = dbClient.transaction: db =>
        db.runSql[Double](sql"""
            WITH retirement_goal AS (
              SELECT id
              FROM goals
              WHERE goals.goal_type = 'retirement' AND goals.user_id = $userId
              LIMIT 1
            ),
            assigned_accounts AS (
              SELECT account_id
              FROM goal_accounts
              JOIN retirement_goal ON retirement_goal.id = goal_accounts.goal_id
            ),
            start_of_month_balances AS (
              -- Get the earliest balance in the current month for each account
              SELECT DISTINCT ON (ab.account_id)
                ab.account_id,
                ab.balance_date AS start_balance_date,
                ab.current_balance AS start_balance
              FROM account_balances ab
              WHERE ab.balance_date >= date_trunc('month', CURRENT_DATE)
                AND ab.account_id IN (SELECT account_id FROM assigned_accounts)
              ORDER BY ab.account_id, ab.balance_date
            ),
            most_recent_balances AS (
              -- Get the most recent balance for each account
              SELECT DISTINCT ON (ab.account_id)
                ab.account_id,
                ab.balance_date AS most_recent_balance_date,
                ab.current_balance AS most_recent_balance
              FROM account_balances ab
              WHERE ab.account_id IN (SELECT account_id FROM assigned_accounts)
              ORDER BY ab.account_id, ab.balance_date DESC
            )
            SELECT
              SUM(mr.most_recent_balance - som.start_balance) AS net_balance_change
            FROM most_recent_balances mr
            JOIN start_of_month_balances som ON mr.account_id = som.account_id
            JOIN accounts ON mr.account_id = accounts.id
            WHERE mr.most_recent_balance_date >= som.start_balance_date
          """)
      result.headOption.getOrElse(0.0)
    .toEither.left.map(e => AppError.DatabaseError(e.getMessage))

  def calculateTargetSavingsAndInvestmentsThisMonth(userId: UUID)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Double] =
    for
      profile <- ProfileRepository.getProfile(userId)
      retirementGoalCurrentBalance <- GoalRepository
        .getAssignedBalanceOnRetirementGoal(userId)
      pv = -(retirementGoalCurrentBalance.abs)
      fv <- getFreedomFutureValueOfCurrentExpensesAtRetirement(
        ExpenseCalculation.Average,
        userId
      )
      years <- calculateYearsToRetirement(userId)
      yearlySavingsTarget = calculatePaymentFromFutureValue(
        pv = pv,
        fv = fv,
        rate = 0.08,
        years = years
      )
    yield yearlySavingsTarget / 12

  def calculateActualSavingsAtRetirement(
      userId: UUID,
      expCalc: ExpenseCalculation
  )(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Double] =
    for
      profile <- ProfileRepository.getProfile(userId)
      annualInterestRate = 0.08
      yearsToRetirement <- calculateYearsToRetirement(userId)
      retirementGoalCurrentBalance <- getAssignedBalanceOnRetirementGoal(userId)
      pv = -(retirementGoalCurrentBalance.abs)
      savings <- calculateActualSavingsThisMonth(userId)
      pmt = -(savings.abs)
    yield calculateFutureValue(
      rate = annualInterestRate,
      years = yearsToRetirement,
      pv = pv,
      pmt = pmt
    )

  def calculateExpectedSavingsAtRetirement(
      exp: ExpenseCalculation,
      userId: UUID
  )(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Double] =
    for
      profile <- ProfileRepository.getProfile(userId)
      yearsToRetirement <- calculateYearsToRetirement(userId)
      freedomFVCurrentExpense <- getFreedomFutureValueOfCurrentExpenses(
        exp,
        userId
      )
      fv = freedomFVCurrentExpense.abs
      rate = 0.02 // 2% annual inflation rate
      pv = freedomFVCurrentExpense.abs
    yield calculateFutureValue(
      rate = rate,
      years = yearsToRetirement,
      pv = pv,
      pmt = 0
    )

  /** How much it costs to maintain the current lifestyle at retirement
    */
  private def getFreedomFutureValueOfCurrentExpenses(
      expCalc: ExpenseCalculation,
      userId: UUID
  )(using dbClient: DbClient.DataSource): Either[AppError, Double] =
    expCalc match
      case ExpenseCalculation.Last12Months =>
        getLast12MonthsInflowOutflow(userId)
      case ExpenseCalculation.Average =>
        getAverageMonthlyInflowOutflow(userId).map(
          _.outflow * 12 / 0.04 // 4% withdrawal rate
        )

  private def getFreedomFutureValueOfCurrentExpensesAtRetirement(
      expCalc: ExpenseCalculation,
      userId: UUID
  )(using dbClient: DbClient.DataSource): Either[AppError, Double] =
    for
      profile <- ProfileRepository.getProfile(userId)
      freedomFutureValueOfCurrentExpenses <-
        getFreedomFutureValueOfCurrentExpenses(expCalc, userId)
      yearsToRetirement <- calculateYearsToRetirement(userId)
    yield calculateFutureValue(
      pv = freedomFutureValueOfCurrentExpenses,
      pmt = 0,
      rate = 0.02, // 2% annual inflation rate
      yearsToRetirement
    )

  def getLast12MonthsInflowOutflow(userId: UUID)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Double] =
    val result =
      Try:
        dbClient.transaction: db =>
          db.runSql[Double](sql"""
            WITH regular_transactions AS (
                SELECT
                    SUM(CASE WHEN amount < 0 THEN ABS(amount) ELSE 0 END) AS total_inflows,
                    SUM(CASE WHEN amount > 0 THEN amount ELSE 0 END) AS total_outflows
                FROM
                    transactions
                    JOIN accounts ON transactions.account_id = accounts.id
                WHERE
                    date >= (CURRENT_DATE - INTERVAL '12 months')
                    AND category NOT IN ('TRANSFER_IN', 'TRANSFER_OUT')
                    AND accounts.user_id = $userId
            ),
            transfer_transactions AS (
                SELECT
                    SUM(CASE WHEN category = 'TRANSFER_IN' THEN amount ELSE -amount END) AS net_transfer
                FROM
                    transactions
                JOIN accounts on transactions.account_id = accounts.id
                WHERE
                    date >= (CURRENT_DATE - INTERVAL '12 months')
                    AND category IN ('TRANSFER_IN', 'TRANSFER_OUT')
                    AND accounts.user_id = $userId
            )
            SELECT
                CASE
                    WHEN tt.net_transfer > 0 THEN rt.total_inflows + tt.net_transfer
                    ELSE rt.total_inflows
                END AS adjusted_inflows,
                CASE
                    WHEN tt.net_transfer < 0 THEN rt.total_outflows - tt.net_transfer
                    ELSE rt.total_outflows
                END AS adjusted_outflows
                        FROM
                            regular_transactions rt, transfer_transactions tt;
          """)

    result.toEither.left
      .map(e => AppError.DatabaseError(e.getMessage))
      .map(rs => rs.headOption.getOrElse(0.0))

  case class AverageMonthlyInflowOutflow(
      inflow: Double,
      outflow: Double
  )

  def getAverageMonthlyInflowOutflow(userId: UUID)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, AverageMonthlyInflowOutflow] =
    val result =
      Try:
        dbClient.transaction: db =>
          db.runSql[(Double, Double)](sql"""
            WITH monthly_regular_transactions AS (
              SELECT
                date_trunc('month', date) AS month,
                sum(
                  CASE WHEN amount < 0 THEN
                    abs(amount)
                  ELSE
                    0
                  END) AS monthly_inflows,
                sum(
                  CASE WHEN amount > 0 THEN
                    amount
                  ELSE
                    0
                  END) AS monthly_outflows
              FROM
                transactions
                JOIN accounts ON transactions.account_id = accounts.id
              WHERE
                date >= (CURRENT_DATE - INTERVAL '12 months')
                AND category NOT IN ('TRANSFER_IN', 'TRANSFER_OUT')
                AND accounts.user_id = $userId
              GROUP BY
                date_trunc('month', date)
            ),
            monthly_transfer_transactions AS (
              SELECT
                date_trunc('month', date) AS month,
                sum(
                  CASE WHEN category = 'TRANSFER_IN' THEN
                    amount
                  ELSE
                    - amount
                  END) AS net_transfer
              FROM
                transactions join accounts on transactions.account_id = accounts.id
              WHERE
                date >= (CURRENT_DATE - INTERVAL '12 months')
                AND category IN ('TRANSFER_IN', 'TRANSFER_OUT')
                AND accounts.user_id = $userId
              GROUP BY
                date_trunc('month', date)
            ),
            adjusted_monthly_totals AS (
              SELECT
                mrt.month,
                CASE WHEN coalesce(mtt.net_transfer, 0) > 0 THEN
                  mrt.monthly_inflows + coalesce(mtt.net_transfer, 0)
                ELSE
                  mrt.monthly_inflows
                END AS adjusted_inflows,
                CASE WHEN coalesce(mtt.net_transfer, 0) < 0 THEN
                  mrt.monthly_outflows - coalesce(mtt.net_transfer, 0)
                ELSE
                  mrt.monthly_outflows
                END AS adjusted_outflows
              FROM
                monthly_regular_transactions mrt
                LEFT JOIN monthly_transfer_transactions mtt ON mrt.month = mtt.month
            )
            SELECT
              avg(adjusted_inflows) AS inflows,
              avg(adjusted_outflows) AS outflows
            FROM
              adjusted_monthly_totals;

          """)

    result.toEither.left
      .map(e => AppError.DatabaseError(e.getMessage))
      .map(rs =>
        rs.headOption
          .map { case (inflow, outflow) =>
            AverageMonthlyInflowOutflow(inflow, outflow)
          }
          .getOrElse(AverageMonthlyInflowOutflow(0.0, 0.0))
      )

  def calculatePeriodFromFutureValue(
      fv: Double,
      pv: Double,
      payment: Double,
      annualInterestRate: Double
  ): Int =
    if (annualInterestRate == 0) then
      // If interest rate is 0, use simple division
      ((fv - pv) / payment).round.toInt
    else
      val absFv = fv.abs
      val absPv = pv.abs
      val absPayment = payment.abs

      val n = math.log(
        (absFv * annualInterestRate + absPayment) /
          (absPv * annualInterestRate + absPayment)
      ) /
        math.log(1 + annualInterestRate)

      n.round.toInt

  def calculateYearsToRetirement(userId: UUID)(using
      dbClient: DbClient.DataSource
  ): Either[AppError, Int] =
    for
      profile <- ProfileRepository.getProfile(userId)
      retirementAge <- profile.retirementAge.toRight(
        AppError.NotFoundError("Retirement age not found")
      )
      age <- profile.age.toRight(AppError.NotFoundError("Age not found"))
    yield retirementAge - age

  /** Calculates the future value of an investment with a given present value,
    * yearly investment, annual interest rate, and number of years.
    *
    * @param presentValue
    *   The initial amount of money.
    * @param yearlyInvestment
    *   The amount of money added to the investment each year.
    * @param annualInterestRate
    *   The annual interest rate as a decimal. Example: 0.08 for 8%
    * @param years
    *   The number of years the investment will grow.
    * @return
    *   The future value of the investment.
    */
  def calculateFutureValue(
      pv: Double,
      pmt: Double,
      rate: Double,
      years: Int
  ): Double =
    val futureValue = pv * math.pow(1 + rate, years) +
      pmt * ((math.pow(
        1 + rate,
        years
      ) - 1) / rate)

    if (pmt < 0 && pv < 0) then -futureValue
    else futureValue

  /** Calculates the payment from the future value of an investment.
    *
    * @param fv
    *   The future value of the investment.
    * @param pv
    *   The present value of the investment.
    * @param rate
    *   The annual interest rate as a decimal. Example: 0.08 for 8%
    * @param years
    *   The number of years the investment will grow.
    * @return
    */
  def calculatePaymentFromFutureValue(
      fv: Double,
      pv: Double,
      rate: Double,
      years: Int
  ): Double =
    // Calculate the payment using the future value formula for annual payments
    val payment =
      if (rate == 0) then
        // If interest rate is 0, use simple division
        (fv - pv) / years
      else
        (fv - pv * math.pow(1 + rate, years)) /
          ((math.pow(1 + rate, years) - 1) / rate)

    // Adjust the sign if both FV and PV are negative
    val adjustedPayment = if (fv < 0 && pv < 0) then -payment else payment

    // Round to 2 decimal places
    BigDecimal(adjustedPayment)
      .setScale(2, BigDecimal.RoundingMode.HALF_UP)
      .toDouble
